addons/website/tools.py

```diff
@@ -187,3 +192,184 @@ def get_base_domain(url, strip_www=False):
+                snippet_name, 'new_page_template_%s', f'new_page_template_{group}_%s'
+            ))
+    create_count += create_missing(create_values)
+
+    # Generate new page snippet templates for specific new page templates within groups
+
+    create_values = []
+    for group in templates:
+        for template_name in templates[group]:
+            for snippet_name in templates[group][template_name]:
+                create_values.append(get_create_vals(
+                    f"Snippet {snippet_name} for new page {group} template {template_name}",
+                    snippet_name, f'new_page_template_{group}_%s', f'new_page_template_{group}_{template_name}_%s'
+                ))
+    create_count += create_missing(create_values)
+
+    if module == 'website':
+        # TODO Find a way to create theme and other module's template patches
+        # Create or update template views per group x key
+        create_values = []
+        for group in templates:
+            for template_name in templates[group]:
+                xmlid = f'{module}.new_page_template_sections_{group}_{template_name}'
+                wrapper = f'%s.new_page_template_{group}_{template_name}_%s'
+                calls = ''.join([
+                    f'<t t-snippet-call="{wrapper % split_key(snippet_key)}"/>'
+                    for snippet_key in templates[group][template_name]
+                ])
+                create_values.append({
+                    'name': f"New page template: {group} #{template_name}",
+                    'type': 'qweb',
+                    'key': xmlid,
+                    'arch': f'<div id="wrap">{calls}</div>',
+                })
+        keys = [values['key'] for values in create_values]
+        existing_primary_templates = env['ir.ui.view'].search_read([('mode', '=', 'primary'), ('key', 'in', keys)], ['key'])
+        existing_primary_template_keys = {data['key']: data['id'] for data in existing_primary_templates}
+        missing_create_values = []
+        update_count = 0
+        for create_value in create_values:
+            if create_value['key'] in existing_primary_template_keys:
+                env['ir.ui.view'].browse(existing_primary_template_keys[create_value['key']]).with_context(no_cow=True).write({
+                    'arch': create_value['arch'],
+                })
+                update_count += 1
+            else:
+                missing_create_values.append(create_value)
+        if missing_create_values:
+            missing_records = env['ir.ui.view'].create(missing_create_values)
+            create_model_data(missing_records)
+            _logger.info('Generated %s primary page templates for %r', len(missing_create_values), module)
+        # Prevent deletion by _process_end
+        for values in create_values:
+            env['ir.model.data']._load_xmlid(values['key'])
+        if update_count:
+            _logger.info('Updated %s primary page templates for %r', update_count, module)
```

Comment:
Can you remind me why we can't create those views directly in xml, as this is not modular but only bound to website anyway?

Reply:
As far as I remember, the initial motivation was because we did want to avoid putting the burden of handling the `primary` templates on the designers - and somehow guarantee that the required tree of template was always correctly defined (without pre-definining 80K+ snippet templates).

~But now, technically, I think we would lack the possibility for themes to define the missing `website.new_page_template_*` snippet templates from their xml files. (Even though they could use their own module namespace to achieve the same thing - but then they would need to handle the `primary` aspect of it too)~

~Oops... that was not really the question~

I think that technically we could define the template views in XML directly.
But if we do that, it will look a bit like a code duplication of those views, because we will still need their definition in the manifest in order to generate the snippets primary template hierarchy...
unless we create the hierarchy by parsing the `t-snippet-call`s from the views instead of looking in the manifest - but doing that sounds like it would require a partial QWeb inheritance browsing capability which seems a bit overkill to me.

Reply:
You are right, I got a bit lost in the PR at the end of the day :D